### PR TITLE
Rasterizer cache refactor v2

### PR DESCRIPTION
**Pull request by GPUCode**
The last PR contained too many changes alreaady so this one intends to finish the job of preparing the rasterizer cache for multiple graphics APIs. The cache is converted into a template which can be specialized by each graphics backend. There's also a series of optimizations that should boost performance in most games, most notably the switch to a page table for surface lookups instead of an interval map, since the operation isn't interval sensitive. This alone shaves about ~1ms of frame time in games with moderate cache usage.

εικόνα
master
εικόνα
PR

Another change is reworking the reinterpretation method to avoid redudant surface lookups that would occur on every surface validation. Instead a new match flag is introduced for reinterpretation which performs a relaxed lookup and lets the validate method decide whether to keep or use the target surface. The functionality is similar to before, while the framework for dimentional reinterpretation is laid out. Notably dimentional reinterpretations with a different formats are now ignored. The only game that does this to my knowledge is Kid Icarus and the result is not used by the game itself. Either way the result of such an operation is never actually useable and most likely the game reusing memory. While this does not completely eliminate all flushes it does significantly boost performance in the game
εικόνα
master
εικόνα
PR

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6479)

### CODE REVIEW

**Comment by GPUCode** [comment]
Tagging this for canary as it fixes a decent chunk of regressions

**Comment by GPUCode** [comment]
Untagging because it conflicts with https://github.com/citra-emu/citra/pull/6456

File: src/video_core/rasterizer_cache/framebuffer_base.cpp [hunk 1]
```diff
@@ -10,9 +10,9 @@ namespace VideoCore {
 
 FramebufferBase::FramebufferBase() = default;
 
-FramebufferBase::FramebufferBase(const Pica::Regs& regs, const SurfaceBase* const color,
-                                 u32 color_level, const SurfaceBase* const depth_stencil,
-                                 u32 depth_level, Common::Rectangle<u32> surfaces_rect) {
+FramebufferBase::FramebufferBase(const Pica::Regs& regs, const SurfaceBase* color, u32 color_level,
```
Comment (FearlessTobi) [inline src/video_core/rasterizer_cache/framebuffer_base.cpp:13]: why is the second const removed here?

File: src/video_core/rasterizer_cache/slot_vector.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
+// SPDX-FileCopyrightText: Copyright 2020 yuzu Emulator Project
```
Comment (FearlessTobi) [inline src/video_core/rasterizer_cache/slot_vector.h:1]: we do not use spdx copyright-headers

File: src/video_core/rasterizer_cache/framebuffer_base.cpp (hunk 1 above)
Comment (GPUCode) [reply src/video_core/rasterizer_cache/framebuffer_base.cpp:13]: The later const was redundant and not consistent with the rest of the codebase (I haven't seen const T* const being used anywhere else)

File: src/video_core/CMakeLists.txt [hunk 3]
```diff
@@ -126,7 +128,7 @@ target_include_directories(video_core PRIVATE ${HOST_SHADERS_INCLUDE})
 create_target_directory_groups(video_core)
 
 target_link_libraries(video_core PUBLIC citra_common citra_core)
-target_link_libraries(video_core PRIVATE glad json-headers dds-ktx nihstro-headers Boost::serialization)
+target_link_libraries(video_core PRIVATE glad tsl::robin_map json-headers dds-ktx nihstro-headers Boost::serialization)
```
Comment (SachinVin) [inline src/video_core/CMakeLists.txt:131]: nit, should be sorted

File: src/video_core/custom_textures/material.cpp [hunk 4]
```diff
@@ -55,6 +55,11 @@ CustomTexture::CustomTexture(Frontend::ImageInterface& image_interface_)
 CustomTexture::~CustomTexture() = default;
 
 void CustomTexture::LoadFromDisk(bool flip_png) {
+    std::scoped_lock lock{decode_mutex};
```
Comment (SachinVin) [inline src/video_core/custom_textures/material.cpp:58]: what is this mutex for ?

File: src/video_core/renderer_opengl/gl_texture_runtime.h [hunk 5]
```diff
@@ -223,4 +211,32 @@ class Framebuffer : public VideoCore::FramebufferBase {
...
+
+    [[nodiscard]] GLuint Handle() const noexcept {
+        return sampler.handle;
+    }
+
+private:
+    OGLSampler sampler;
+};
+
+struct Traits {
```
Comment (SachinVin) [inline src/video_core/renderer_opengl/gl_texture_runtime.h:233]: could've been a namespace ?

File: src/video_core/renderer_opengl/gl_texture_runtime.cpp [hunk 6]
```diff
@@ -692,4 +682,30 @@ Framebuffer::Framebuffer(TextureRuntime& runtime, Surface* const color, u32 colo
 
 Framebuffer::~Framebuffer() = default;
 
+Sampler::Sampler(TextureRuntime& runtime, VideoCore::SamplerParams params) {
```
Comment (SachinVin) [inline src/video_core/renderer_opengl/gl_texture_runtime.cpp:685]: runtime is unused?

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 7]
```diff
@@ -538,22 +527,14 @@ void RasterizerOpenGL::SyncTextureUnits(const Framebuffer& framebuffer) {
...
-        if (!surface) {
-            // Can occur when texture addr is null or its memory is unmapped/invalid
-            // HACK: In this case, the correct behaviour for the PICA is to use the last
-            // rendered colour. But because this would be impractical to implement, the
-            // next best alternative is to use a clear texture, essentially skipping
-            // the geometry in question.
-            // For example: a bug in Pokemon X/Y causes NULL-texture squares to be drawn
-            // on the male character's face, which in the OpenGL default appear black.
-            state.texture_units[texture_index].texture_2d = default_texture;
-        } else if (!IsFeedbackLoop(texture_index, framebuffer, *surface)) {
```
Comment (SachinVin) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp:554]: Where is this handled ?

File: src/video_core/renderer_opengl/gl_blit_helper.cpp [hunk 8]
```diff
@@ -49,43 +53,110 @@ OGLProgram CreateProgram(std::string_view frag) {
...
+        surface.type == SurfaceType::Depth || surface.type == SurfaceType::DepthStencil;
+    if (filter == Settings::TextureFilter::None || is_depth) {
         return false;
     }
-    // Avoid filtering for mipmaps as the result often looks terrible.
     if (blit.src_level != 0) {
         return true;
     }
 
     const OpenGLState prev_state = OpenGLState::GetCurState();
```
Comment (SachinVin) [inline src/video_core/renderer_opengl/gl_blit_helper.cpp:156]: IMO it makes sense to move this into Filter* instead.

File: src/video_core/renderer_opengl/gl_blit_helper.cpp [hunk 9]
```diff
@@ -49,43 +53,110 @@ OGLProgram CreateProgram(std::string_view frag) {
...
+        LOG_INFO(Render_OpenGL,
+                 "Texture views are unsupported, reinterpretation will do intermediate copy");
+        temp_tex.Create();
+        use_texture_view = false;
+    }
 }
 
 BlitHelper::~BlitHelper() = default;
 
+bool BlitHelper::ConvertDS24S8ToRGBA8(Surface& source, Surface& dest,
```
Comment (SachinVin) [inline src/video_core/renderer_opengl/gl_blit_helper.cpp:84]: Was the PBO path removed, afaik Intel drivers are buggy when you use this

File: src/video_core/rasterizer_cache/rasterizer_cache.h
Comment (SachinVin) [inline src/video_core/rasterizer_cache/rasterizer_cache.h:1]: Can you move stuff to the cpp file, this is huge even for a .cpp file let alone a header file.

File: src/video_core/rasterizer_cache/rasterizer_cache.h [hunk 10]
```diff
@@ -2,190 +2,1402 @@
 // Licensed under GPLv2 or any later version
 // Refer to the license.txt file included.
 
-#pragma once
```
Comment (SachinVin) [inline src/video_core/rasterizer_cache/rasterizer_cache.h:5]: why was this removed ?

File: src/video_core/rasterizer_cache/rasterizer_cache.h [hunk 11]
```diff
@@ -2,190 +2,1402 @@
...
+}
+
+template <class T>
+typename RasterizerCache<T>::SurfaceRect_Tuple RasterizerCache<T>::GetTexCopySurface(
+    const SurfaceParams& params) {
+    Common::Rectangle<u32> rect{};
+
+    SurfaceId match_id = FindMatch<MatchFlags::TexCopy>(params, ScaleMatch::Ignore);
+
+    if (match_id) {
```
Comment (SachinVin) [inline src/video_core/rasterizer_cache/rasterizer_cache.h:720]: consider returning std::optional 
```suggestion
    if (!match_id) {
        return std::nullopt;
    }
```

File: src/video_core/rasterizer_cache/surface_base.h [hunk 12]
```diff
@@ -30,15 +39,23 @@ class SurfaceBase : public SurfaceParams {
...
 
+    /// Returns the internal surface extent.
+    Extent RealExtent(bool scaled = true);
+
     /// Returns true if the surface contains a custom material with a normal map.
     bool HasNormalMap() const noexcept;
 
+    bool Overlaps(PAddr overlap_addr, size_t overlap_size) const noexcept {
+        const PAddr overlap_end = overlap_addr + static_cast<PAddr>(overlap_size);
+        return addr < overlap_end && overlap_addr < end;
```
Comment (SachinVin) [inline src/video_core/rasterizer_cache/surface_base.h:50]: is `end - addr` always equal to `overlap_size`?

File: src/video_core/custom_textures/material.cpp (hunk 4 above)
Comment (GPUCode) [reply src/video_core/custom_textures/material.cpp:58]: This PR implement support for mapping a texture to multiple hashes. This means that two different materials might point to the same texture. That's a problem when using async decode as two threads might attempt to load the same texture at once. The mutex prevents that happening.

File: src/video_core/renderer_opengl/gl_texture_runtime.h (hunk 5 above)
Comment (GPUCode) [reply src/video_core/renderer_opengl/gl_texture_runtime.h:233]: How? I can't see any other way of passing type information to the rasterizer cache template

File: src/video_core/renderer_opengl/gl_texture_runtime.cpp (hunk 6 above)
Comment (GPUCode) [reply src/video_core/renderer_opengl/gl_texture_runtime.cpp:685]: It's not used here (well it was used before the lod bias patch) but vulkan requires it and the interface must be same because template class and all

File: src/video_core/renderer_opengl/gl_blit_helper.cpp [hunk 13]
```diff
@@ -49,43 +53,110 @@ OGLProgram CreateProgram(std::string_view frag) {
...
+      d24s8_to_rgba8{CreateProgram(HostShaders::D24S8_TO_RGBA8_FRAG)},
+      rgba4_to_rgb5a1{CreateProgram(HostShaders::RGBA4_TO_RGB5A1_FRAG)} {
     vao.Create();
-    filter_fbo.Create();
+    draw_fbo.Create();
     state.draw.vertex_array = vao.handle;
     for (u32 i = 0; i < 3; i++) {
         state.texture_units[i].sampler = i == 2 ? nearest_sampler.handle : linear_sampler.handle;
     }
+    if (driver.IsOpenGLES()) {
```
[the remark this answers is not in the document]
Comment (GPUCode) [reply src/video_core/renderer_opengl/gl_blit_helper.cpp:74]: Texture views are used here for getting the stencil view of a depth stencil texture without an intermediate copy, which doesn't violate the spec. I've confirmed this working on the new AMD drivers as well

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 14]
```diff
@@ -726,24 +710,20 @@ void RasterizerOpenGL::NotifyFixedFunctionPicaRegisterChanged(u32 id) {
 }
 
 void RasterizerOpenGL::FlushAll() {
-    MICROPROFILE_SCOPE(OpenGL_CacheManagement);
```
[the remark this answers is not in the document]
Comment (GPUCode) [reply src/video_core/renderer_opengl/gl_rasterizer.cpp:729]: Because it has been moved to the rasterizer cache

File: src/video_core/renderer_opengl/gl_rasterizer.cpp (hunk 7 above)
Comment (GPUCode) [reply src/video_core/renderer_opengl/gl_rasterizer.cpp:554]: [Here](https://github.com/GPUCode/citra/blob/f40f49cfeb548215d6c72e803cbc2073736b3a4d/src/video_core/rasterizer_cache/rasterizer_cache.h#L471) inside GetTextureSurface. Also while opengl can bind 0 to a texture unit when it returns null, vulkan cannot so I believe its better there as vulkan won't need to manage its own default textures

File: src/video_core/rasterizer_cache/rasterizer_cache.h (hunk 10 above)
Comment (GPUCode) [reply src/video_core/rasterizer_cache/rasterizer_cache.h:5]: Probably an oversight. Though this header file isn't meant to be included publicly, instead it's included in gl_rasterizer_cache.cpp for specialization

File: src/video_core/renderer_opengl/gl_blit_helper.cpp (hunk 8 above)
Comment (GPUCode) [reply src/video_core/renderer_opengl/gl_blit_helper.cpp:156]: The opengl state? Wouldn't that unnecessary duplicate it on every filter function?

File: src/video_core/rasterizer_cache/rasterizer_cache.h (hunk 11 above)
Comment (GPUCode) [reply src/video_core/rasterizer_cache/rasterizer_cache.h:720]: The only place this function is used is in AccelerateTextureCopy which [checks](https://github.com/GPUCode/citra/blob/f40f49cfeb548215d6c72e803cbc2073736b3a4d/src/video_core/rasterizer_cache/rasterizer_cache.h#L137) the result immediately after so std::optional would be superflous here imo

File: src/video_core/renderer_opengl/gl_blit_helper.cpp (hunk 9 above)
Comment (GPUCode) [reply src/video_core/renderer_opengl/gl_blit_helper.cpp:84]: Uh the PBO path was removed way back in August because we don't support those legacy intel [drivers](https://github.com/citra-emu/citra/blob/17ad594a62915177bf9227065c4349076db724df/src/video_core/renderer_opengl/gl_format_reinterpreter.cpp#L375) I think. Also we now require all extensions for the shader path. But if it still has issues I can add it back

File: src/video_core/rasterizer_cache/rasterizer_cache.h
Comment (GPUCode) [reply src/video_core/rasterizer_cache/rasterizer_cache.h:1]: Tried that but ended up with a lot of linking errors. Do you have any idea? Spliting the class in two is not feasible since 90% of functions depend on the template state in one way or another. Also I'd argue this isn't that big of an issue because the header is only included in gl_rasterizer_cache.cpp to explicitly define the template.

File: src/video_core/renderer_opengl/gl_blit_helper.cpp (hunk 9 above)
Comment (SachinVin) [reply src/video_core/renderer_opengl/gl_blit_helper.cpp:84]: I think even 27.xx also are broken, It's been a while so I might be wrong though

File: src/video_core/renderer_opengl/gl_blit_helper.cpp (hunk 8 above)
Comment (SachinVin) [reply src/video_core/renderer_opengl/gl_blit_helper.cpp:156]: Since state is applied in the Filter* functions, I think it would make sense to save and revert it there

File: src/video_core/renderer_opengl/gl_rasterizer.cpp (hunk 7 above)
Comment (SachinVin) [reply src/video_core/renderer_opengl/gl_rasterizer.cpp:554]: Got it, thanks

File: src/video_core/renderer_opengl/gl_texture_runtime.cpp (hunk 6 above)
Comment (SachinVin) [reply src/video_core/renderer_opengl/gl_texture_runtime.cpp:685]:
```suggestion
Sampler::Sampler(TextureRuntime&, VideoCore::SamplerParams params) {
```
or
```suggestion
Sampler::Sampler([[maybe_unused]] TextureRuntime& runtime, VideoCore::SamplerParams params) {
```

File: src/video_core/renderer_opengl/gl_texture_runtime.h (hunk 5 above)
Comment (SachinVin) [reply src/video_core/renderer_opengl/gl_texture_runtime.h:233]: https://github.com/yuzu-emu/yuzu/blob/master/src/video_core/texture_cache/texture_cache_base.h#L110 this is how yuzu does it

Comment (GPUCode) [reply src/video_core/renderer_opengl/gl_texture_runtime.h:233]: I'm doing the same thing as yuzu though

Comment (SachinVin) [reply src/video_core/renderer_opengl/gl_texture_runtime.h:233]: wouldn't `using RasterizerCache = VideoCore::RasterizerCache<OpenGL>;` work ?

File: src/video_core/rasterizer_cache/slot_vector.h
Comment (SachinVin) [inline src/video_core/rasterizer_cache/slot_vector.h:1]: Maybe makes sense to move it to common ?

File: src/video_core/custom_textures/material.cpp (hunk 4 above)
Comment (GPUCode) [reply src/video_core/custom_textures/material.cpp:58]: They can be considered the same really, CustomTexture = a png file in the load directory

Comment (SachinVin) [reply src/video_core/custom_textures/material.cpp:58]: Yes but if it was the file, then it would be read-only, and doesn't need a mutex.

Comment (GPUCode) [reply src/video_core/custom_textures/material.cpp:58]: Imagine the following scenario. The game has 2 ropes with slightly different textures next to each other. A pack creator might want to use the same texture for both of these, so they will map rope.png to hash1 and hash2. This translates to 2 materials in the code refering to the same CustomTexture object. Then when enabling custom textures with async decode, the game tries to use the first rope texture. The cache sees its not decoded yet so it queues the decode in the thread pool. Next draw it uses the second rope texture, cache sees the material isn't loaded and queues it for loading. Thus you end up in a situation where two threads might call LoadFromDisk at the same time. I know it's hightly unlikely but a possibility nonetheless.

Comment (SachinVin) [reply src/video_core/custom_textures/material.cpp:58]: So i assume the CustomTexture is created for each file rather than each hash, is that right?

Comment (GPUCode) [reply src/video_core/custom_textures/material.cpp:58]: Yes

Comment (SachinVin) [reply src/video_core/custom_textures/material.cpp:58]: Got it, thanks
